s3transfer/download.py

```diff
@@ -475,3 +558,38 @@ def __init__(self, transfer_coordinator, main_kwargs=None,
 
     def _main(self):
         pass
+
+
+class DeferQueue(object):
+    """IO queue that defers write requests until they are queued sequentially.
+
+    This class is used to track IO data for a *single* fileobj.
+
+    You can send data to this queue, and it will defer any IO write requests
+    until it has the next contiguous block available (starting at 0).
+
+    """
+    def __init__(self):
+        self._writes = []
+        self._next_offset = 0
+
+    def request_writes(self, offset, data):
+        """Request any available writes given new incoming data.
+
+        You call this method by providing new data along with the
+        offset associated with the data.  If that new data unlocks
+        any contiguous writes that can now be submitted, this
+        method will return all applicable writes.
+
+        This is done with 1 method call so you don't have to
+        make two method calls (put(), get()) which acquires a lock
+        each method call.
+
+        """
+        writes = []
+        heapq.heappush(self._writes, (offset, data))
+        while self._writes and self._writes[0][0] == self._next_offset:
```

Comment:
What I had in mind was:

* Ignore/discard any writes that are less than the `_next_offset`.
* Ignore any duplicate entries we receive for pending writes.  Based on the atomicity of `request_writes` and the guaranteed order of queued IO within a given GetObjectTask I don't actually think this is possible in practice, but I'd rather not have this class rely on that assumption for correctness.

I think that in order to address the second point I need to add an auxiliary structure to keep track of what offsets are pending (probably just a set of integers of offsets).